### PR TITLE
IGNITE-15794 Implicit call attributes for java service.

### PR DESCRIPTION
**Pull request by xtern**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceCallContextImpl.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    public ServiceCallContextImpl(Map<String, byte[]> attrs) {
+        this.attrs = Collections.unmodifiableMap(attrs);
+    }
+
+    /**
+     * Constructs an immutable context.
+     *
+     * @param callCtx Service call context.
+     */
+    public ServiceCallContextImpl(ServiceCallContext callCtx) {
```
Comment (ivandasch) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceCallContextImpl.java:55]: I suppose that it is a really bad idea.

Suppose that someone pass another implementation of ServiceCallContext?

Is it ok to fail in runtime in this case?

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java [hunk 2]
```diff
@@ -374,8 +376,34 @@ public IgniteServicesImpl(GridKernalContext ctx, ClusterGroupAdapter prj, boolea
...
     /** {@inheritDoc} */
-    @Override public <T> T serviceProxy(final String name, final Class<? super T> svcItf, final boolean sticky,
-        final long timeout) throws IgniteException {
+    @Override public <T> T serviceProxy(
+        final String name,
+        final Class<? super T> svcItf,
+        final boolean sticky,
+        final long timeout
+    ) throws IgniteException {
+        return (T) serviceProxy(name, svcItf, sticky, (Supplier<ServiceCallContext>)null, timeout);
```
Comment (ivandasch) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java:385]: Remove please extra space after cast

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java [hunk 3]
```diff
@@ -374,8 +376,34 @@ public IgniteServicesImpl(GridKernalContext ctx, ClusterGroupAdapter prj, boolea
...
+
+    /** {@inheritDoc} */
+    @Override public <T> T serviceProxy(
+        final String name,
+        final Class<? super T> svcItf,
+        final boolean sticky,
+        @Nullable ServiceCallContext callCtx,
+        final long timeout
+    ) throws IgniteException {
+        return (T) serviceProxy(name, svcItf, sticky, callCtx != null ? callCtx::copy : null, timeout);
```
Comment (ivandasch) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java:396]: And the same

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProxy.java [hunk 4]
```diff
@@ -113,14 +116,16 @@
...
      * @param ctx Context.
      */
     public GridServiceProxy(ClusterGroup prj,
         String name,
         Class<? super T> svc,
         boolean sticky,
         long timeout,
-        GridKernalContext ctx)
+        GridKernalContext ctx,
+        @Nullable Supplier<ServiceCallContext> callCtxProvider)
```
Comment (ivandasch) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProxy.java:128]: Please move bracket to new line

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java (hunk 2 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java]: Done, thanks

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java (hunk 3 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteServicesImpl.java]: Done, thanks

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProxy.java (hunk 4 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProxy.java]: Done, thanks

**Review by ivandasch — APPROVED** [review]

**Pull request closed without merging**